lib/msal-node/src/client/ClientApplication.ts

```diff
@@ -279,33 +282,132 @@ export abstract class ClientApplication {
             validRequest.forceRefresh
         );
+
         try {
-            const silentFlowClientConfig =
+            const clientConfiguration =
                 await this.buildOauthClientConfiguration(
                     validRequest.authority,
                     validRequest.correlationId,
                     validRequest.redirectUri || "",
                     serverTelemetryManager,
                     undefined,
-                    request.azureCloudOptions
+                    validRequest.azureCloudOptions
                 );
-            const silentFlowClient = new SilentFlowClient(
-                silentFlowClientConfig
-            );
+            const silentFlowClient = new SilentFlowClient(clientConfiguration);
             this.logger.verbose(
                 "Silent flow client created",
                 validRequest.correlationId
             );
-            return await silentFlowClient.acquireToken(validRequest);
+
+            try {
+                return await this.acquireCachedToken(
+                    validRequest,
+                    silentFlowClient,
+                    clientConfiguration
+                );
+            } catch (e) {
+                try {
+                    // If persistence exists, overwrite the in-memory cache with the persistent cache and search for the cached token again
+                    const persistence = this.config.cache.cachePlugin;
+                    if (persistence) {
+                        this.logger.info(
+                            "Overwriting in-memory cache with persistent cache"
+                        );
+                        this.storage.clear();
+                        const cacheContext = new TokenCacheContext(
+                            this.tokenCache,
+                            false
+                        );
+                        await persistence.beforeCacheAccess(cacheContext);
+                        const cacheSnapshot =
+                            this.tokenCache.getCacheSnapshot();
+                        this.storage.setCache(cacheSnapshot);
+                        await persistence.afterCacheAccess(cacheContext);
+
+                        this.logger.info("Searching again for a valid token");
+                        return await this.acquireCachedToken(
+                            validRequest,
+                            silentFlowClient,
+                            clientConfiguration
+                        );
+                    }
+                } catch {
+                    if (
+                        e instanceof ClientAuthError &&
+                        e.errorCode ===
+                            ClientAuthErrorCodes.tokenRefreshRequired
+                    ) {
+                        const refreshTokenClient = new RefreshTokenClient(
+                            clientConfiguration
+                        );
+                        return refreshTokenClient.acquireTokenByRefreshToken(
```

Comment:
Does this need it's own try/catch or is it covered by the catch on line 369? Same question for line 360.

Reply:
The refactor looks good, thanks.